Optimization: prevent empty COMMIT calls

chaffeqa opened the pull request:
Prevents `BEGIN COMMIT` empty database calls after each rails request if the session did not change.

Don't know why the `write_session` is called in our app every request, but monkey patching using this logic prevents an empty write at the end of each request.

chaffeqa committed:
Optimization: prevent empty COMMIT calls

Prevents `BEGIN COMMIT` empty database calls after each rails request if the session did not change.

Don't know why the `write_session` is called in our app every request, but monkey patching using this logic prevents an empty write at the end of each request.

sikachu reviewed (changes requested):
Thank you for the PR. Please see my inline comment below.

chaffeqa committed:
Update active_record_store.rb

chaffeqa commented on lib/action_dispatch/session/active_record_store.rb:
```diff
@@ -78,7 +78,7 @@ def write_session(request, sid, session_data, options)
           record = get_session_model(request, sid)
           record.data = session_data
-          return false unless record.save
+          return false if record.changed? and not record.save
```
done! sorry for looooong delay

h0jeZvgoxFepBQ2C commented:
Can this be merged?

preethi29 commented:
any update on this? this seems to be useful.

Also the store seems to record empty session on every request without valid session. Is there a way to optimise it other than overriding `save`?

rafaelfranca committed:
Add tests to make sure an empty save isn't made

And fix the code to actually make this test pass.

We now track changes to the session data and only save if the session
has changed or is a new record. This prevents unnecessary database
writes when the session data hasn't changed.
